src/cloud-api-adaptor/test/e2e/docker_test.go

```diff
@@ -0,0 +1,110 @@
+
+package e2e
+
+import (
+	"testing"
+
+	_ "github.com/confidential-containers/cloud-api-adaptor/src/cloud-api-adaptor/test/provisioner/docker"
+)
+
+func TestDockerCreateSimplePod(t *testing.T) {
+	assert := DockerAssert{}
+	DoTestCreateSimplePod(t, testEnv, assert)
+}
+
+func TestDockerCreatePodWithConfigMap(t *testing.T) {
+	SkipTestOnCI(t)
+	assert := DockerAssert{}
+	DoTestCreatePodWithConfigMap(t, testEnv, assert)
+}
+
+func TestDockerCreatePodWithSecret(t *testing.T) {
+	assert := DockerAssert{}
+	DoTestCreatePodWithSecret(t, testEnv, assert)
+}
+
+func TestDockerCreatePeerPodContainerWithExternalIPAccess(t *testing.T) {
+	SkipTestOnCI(t)
+	assert := DockerAssert{}
+	DoTestCreatePeerPodContainerWithExternalIPAccess(t, testEnv, assert)
+
+}
+
+func TestDockerCreatePeerPodWithJob(t *testing.T) {
+	assert := DockerAssert{}
+	DoTestCreatePeerPodWithJob(t, testEnv, assert)
+}
+
+func TestDockerCreatePeerPodAndCheckUserLogs(t *testing.T) {
+	assert := DockerAssert{}
+	DoTestCreatePeerPodAndCheckUserLogs(t, testEnv, assert)
+}
+
+func TestDockerCreatePeerPodAndCheckWorkDirLogs(t *testing.T) {
+	assert := DockerAssert{}
+	DoTestCreatePeerPodAndCheckWorkDirLogs(t, testEnv, assert)
+}
+
+func TestDockerCreatePeerPodAndCheckEnvVariableLogsWithImageOnly(t *testing.T) {
+	// This test is causing issues on CI with instability, so skip until we can resolve this.
+	// See https://github.com/confidential-containers/cloud-api-adaptor/issues/1831
+	SkipTestOnCI(t)
+	assert := DockerAssert{}
+	DoTestCreatePeerPodAndCheckEnvVariableLogsWithImageOnly(t, testEnv, assert)
+}
+
+func TestDockerCreatePeerPodAndCheckEnvVariableLogsWithDeploymentOnly(t *testing.T) {
+	assert := DockerAssert{}
+	DoTestCreatePeerPodAndCheckEnvVariableLogsWithDeploymentOnly(t, testEnv, assert)
+}
+
+func TestDockerCreatePeerPodAndCheckEnvVariableLogsWithImageAndDeployment(t *testing.T) {
+	// This test is causing issues on CI with instability, so skip until we can resolve this.
+	// See https://github.com/confidential-containers/cloud-api-adaptor/issues/1831
+	assert := DockerAssert{}
+	DoTestCreatePeerPodAndCheckEnvVariableLogsWithImageAndDeployment(t, testEnv, assert)
+}
+
+func TestDockerCreateNginxDeployment(t *testing.T) {
+	assert := DockerAssert{}
+	DoTestNginxDeployment(t, testEnv, assert)
+}
+
+/*
+Failing due to issues will pulling image (ErrImagePull)
+func TestDockerCreatePeerPodWithLargeImage(t *testing.T) {
+	assert := DockerAssert{}
+	DoTestCreatePeerPodWithLargeImage(t, testEnv, assert)
+}
+*/
+
+func TestDockerDeletePod(t *testing.T) {
+	assert := DockerAssert{}
+	DoTestDeleteSimplePod(t, testEnv, assert)
+}
+
+func TestDockerPodToServiceCommunication(t *testing.T) {
+	assert := DockerAssert{}
+	DoTestPodToServiceCommunication(t, testEnv, assert)
+}
+
+func TestDockerPodsMTLSCommunication(t *testing.T) {
+	assert := DockerAssert{}
+	DoTestPodsMTLSCommunication(t, testEnv, assert)
+}
+
+func TestDockerKbsKeyRelease(t *testing.T) {
```

Comment:
Did you test with the KBS deployed? I just followed your command, so didn't and I wonder if there are any issues with the kbs in kind and docker?

Reply:
I set up the KBS and it failed:
```
time="2024-07-01T10:04:57-07:00" level=info msg="Do test kbs key release"
=== RUN   TestDockerKbsKeyRelease/KbsKeyReleasePod_test
    assessment_runner.go:265: Waiting for containers in pod: busybox-wget are ready
=== RUN   TestDockerKbsKeyRelease/KbsKeyReleasePod_test/Kbs_key_release_is_successful
    assessment_runner.go:415: Output when execute test commands:command terminated with exit code 1
    assessment_runner.go:417: command terminated with exit code 1
time="2024-07-01T10:05:19-07:00" level=info msg="Deleting pod busybox-wget..."
time="2024-07-01T10:05:24-07:00" level=info msg="Pod busybox-wget has been successfully deleted within 60s"
--- FAIL: TestDockerKbsKeyRelease (53.74s)
    --- PASS: TestDockerKbsKeyRelease/DoTestKbsKeyReleaseForFailure_test (26.31s)
        --- PASS: TestDockerKbsKeyRelease/DoTestKbsKeyReleaseForFailure_test/Kbs_key_release_is_failed (6.26s)
    --- FAIL: TestDockerKbsKeyRelease/KbsKeyReleasePod_test (27.43s)
        --- FAIL: TestDockerKbsKeyRelease/KbsKeyReleasePod_test/Kbs_key_release_is_successful (7.38s)
FAIL
```

Could have been user error though

Reply:
I didn't test KBS

Reply:
ok. Do you think it's worth leaving out the KBS test from `docker_test.go` until we've been able to get it working? We can create a separate issue to track that maybe?

Reply:
Let me try once. There doesn't seem to be anything provider specific for the KBS sample client tests.

Reply:
I tried again with a brand new cluster and go the same failure, there isn't any trace that gives more info on the problem though.

Reply:
sadly, the CDH logs alone don't tell us much here, we'd need to see the AA logs, and probably the KBS logs

Reply:
So when running `curl -LO http://127.0.0.1:8006/cdh/resource/reponame/workload_key/key.bin` from within the pod, I don't see any request reaching the KBS.  

However, from the pod I can reach the KBS directly 
```
curl -L http://172.20.0.2:30764
```

Also `curl -L -vv http://127.0.0.1:8006` works indicating that connection to CDH is working

Reply:
ASR logs
```
Jul 02 11:48:11 36d0845e272d systemd[1]: Started api-server-rest.service - CDH API Server Rest Service.
Jul 02 11:48:11 36d0845e272d api-server-rest[167]: Starting API server on 127.0.0.1:8006 with features all
Jul 02 11:48:11 36d0845e272d api-server-rest[167]: API Server listening on http://127.0.0.1:8006
Jul 02 11:53:45 36d0845e272d api-server-rest[167]: root_path /cdh, url_path /resource/reponame/workload_key/key.bin
Jul 02 11:54:30 36d0845e272d api-server-rest[167]: root_path /cdh, url_path /resource
Jul 02 12:00:48 36d0845e272d api-server-rest[167]: root_path /cdh, url_path /resource/reponame/workload_key/key.bin
Jul 02 12:01:06 36d0845e272d api-server-rest[167]: root_path /cdh, url_path /resource
Jul 02 12:01:29 36d0845e272d api-server-rest[167]: root_path /cdh, url_path /resource
Jul 02 12:01:37 36d0845e272d api-server-rest[167]: root_path /cdh, url_path /resource
Jul 02 12:45:45 36d0845e272d api-server-rest[167]: root_path /cdh, url_path /resource/reponame/workload_key/key.bin
```

So looks like for some reason CDH is not sending the request to KBS.

Reply:
so, if `RUST_LOG=debug` is set, we should see reasons why the connection failed in the AA and CDH logs. I think it's AA that is facilitating the token exchange with KBS first. you can try `curl "http://127.0.0.1:8006/aa/token?token_type=kbs"`, it shouldn't matter whether the pod is able to reach KBS (the pod only talks to the ASR endpoint) but the namespace in which the CDH and AA processes run

Reply:
> So looks like for some reason CDH is not sending the request to KBS.

It works like this, afaik:

ASR => CDH =get_token()=> AA => TokenKBS

once CDH has a token:
ASR => CDH => ResourceKBS

Reply:
ah yes.. right.. let me check

Reply:
@mkulke I was not using the latest build of CAA during my test and aa.toml was not created. That was the reason :-( 

Now I'm able to retrieve secret from KBS with latest CAA 
```
kubectl exec -it hello-world -- bash

[root@hello-world /]# curl -LO http://127.0.0.1:8006/cdh/resource/reponame/workload_key/key.bin
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    34  100    34    0     0    124      0 --:--:-- --:--:-- --:--:--   124
[root@hello-world /]# cat key.bin
This is my cluster name: peer-pods[root@hello-world /]#
```